motionplan/tpSpaceRRT.go

```diff
@@ -0,0 +1,435 @@
+//go:build !windows
+
+package motionplan
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"math/rand"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/motionplan/tpspace"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+var (
+	defaultGoalCheck = 5
+	defaultAutoBB    = 1. // Automatic bounding box on driveable area as a multiple of start-goal distance
+
+	// whether to add intermediate waypoints.
+	defaultAddInt = false
+	// Add a subnode every this many mm along a valid trajectory. Large values run faster, small gives better paths
```

Comment:
Why does small give better paths? I would expect the path to be the same whether or not we add intermediate nodes. And I would expect large values to give us more opportunity to utilize controls on the base.

Reply:
These nodes are not related to controls on the base; this is purely to help us build the RRT tree.

As for why these yield better paths: An example.

Let the start position be (0,0), the goal be (1,2500), and our simulated path length be 5000. Let's say we can't go directly from the start to the goal for some reason.

So we pick a random point and extend from (0,0) towards it. Let's randomly pick (0,9999). Since we simulate  paths of len 5000, the closest trajectory point is (0,5000) and so that is added to the RRT tree.

If defaultAddInt is false, then the RRT tree is now just (0,0) and (0,5000), and so if the goal is reachable from that second point, the path would go through that second node before doubling back for a total length of ~7500.

If defaultAddInt is true, then depending on this distance value, if it were say 2000, then we would add nodes at (0,2000), (0,4000), and (0,5000), and each of those would have an opportunity to be extended from to our goal at (1,2500), yielding more opportunities for a shorter path since we're no longer required to go all the way to (0,5000).

If the distance value is smaller still, say 50 instead of 2000, then we add even more nodes and have even more opportunities to get still shorter paths by branching off closer to the optimal point. The tradeoff being that more nodes to check takes longer.

Reply:
Ah, that makes sense, thank you for explaining. I still think there's a risk that if the paths are too short, then that limits our ability to implement controls in the base. But this could be mitigated by removing intermediate waypoints that are along the same trajectory at the end.